### ISSUE TITLE
Validate/customise OT tables for TTF

### ISSUE THREAD

**Issue by jublo** [source-foundry/meta#18]

**Comment by chrissimpkins** [comment]
Validations:

- [ ] head table
- [ ] hhea table
- [ ] maxp table
- [ ] OS/2 table
- [ ] name table
- [ ] post table
- [ ] TTFA table (ttfautohint)

**Comment by chrissimpkins** [comment]
Adobe features file spec:

https://www.adobe.com/devnet/opentype/afdko/topic_feature_file_syntax.html

**Comment by jublo** [comment]
Head table, checkSumAdjustment attribute: will always be different.

http://stackoverflow.com/questions/25727592/what-is-checksumadjustment-in-ttf-otf-head-table-used-for

**Comment by jublo** [comment]
More references to the HEAD table: https://www.microsoft.com/typography/otspec/head.htm

**Comment by jublo** [comment]
Setting unitsPerEm to 2047 results in the HEAD table having unitsPerEm set to 2045. 2048 (current state) turns to 2049. @chrissimpkins

**Comment by jublo** [comment]
Changes to UFO/fontinfo.plist HEAD `flags` do not carry over to neither OTF nor TTF. Why does makeotf ignore them?

**Comment by jublo** [comment]
openTypeHheaDescender: Should be set to the glyph's lowest cellar depth (descender).  Carries over from features.fea file to both OTF and TTF!

**Comment by jublo** [comment]
@chrissimpkins HHEA is fine. 3.000-dev and 2.019 differ since 3.000 is ASCII only, but that's okay.

**Comment by jublo** [comment]
@chrissimpkins MAXP is fine. Contains maximum values based on included glyphs (maximum number of contours, etc.)

**Comment by jublo** [comment]
OS/2 yStrikeoutPosition is set correctly in fontinfo.plist, but doesnt take over and is thus wrong in built TTF/OTF.

**Comment by jublo** [comment]
Summary of comparison: Most settings that differ between 2.018 release and 3.000-dev originate from the different glyph set and can thus be safely ignored. However, some settings, such as the superscript height, various flags (`isFixedPitch` and more) are ignored from the UFO fontinfo.plist and/or features.fea.template. Assumingly they are lost during AFDKO’s conversion from UFO to T1 format (using `tx`).

The only approaches to fix these and get the font table entries we want are:

- Either patching the values after generating OTF binary, before converting it to TTF, so the OTF patches carry over to TTF (needs testing if they do),
- or patching the values after generating both files, in both binaries.

Patching can take place in two variants:

- Either by dumping to XML using AFDKO `ttx` tool, editing the XML values and converting back (which is what I have done and it works, the `font-tables` tool report on the generated TTF file looks much better than on the raw `make desktop` build),
- Or by having a Python script (which I can’t code).

Working on a proof-of-concept.

**Comment by chrissimpkins** [comment]
> Setting unitsPerEm to 2047 results in the HEAD table having unitsPerEm set to 2045. 2048 (current state) turns to 2049 

I'm not certain why this is happening JM.  I suspect that there is some rounding issue that we are not seeing here.

**Comment by jublo** [comment]
May come in handy if going the TTX approach: http://www.lbreyer.com/xml-sed_man.html

**Comment by chrissimpkins** [comment]
I agree with the post-build OT table modification approach for anything that we cannot modify pre-build.  Let me look into the outfile write capabilities of the Python `fontTools` library (that I used for the font-tables tool).  We will need to make sure that this round trip does not modify other things...  It is really difficult to imagine why no one has put together a single tool capable of taking us from start to release here.  It's as if there are not thousands of fonts out there...

**Comment by jublo** [comment]
@chrissimpkins The TTX roundtrip to XML and back doesnt change anything.

**Comment by chrissimpkins** [comment]
@jublo It looks like most of the incorrect information can safely be reused for each release version (i.e. there is no need to update the replaced values that we initially develop). Is that your opinion as well?

**Comment by jublo** [comment]
@chrissimpkins I don’t understand. Can you please give an example?

**Comment by chrissimpkins** [comment]
Most of those incorrect values will be replaced with correct values that will persist irrespective of version. Is that your understanding as well based upon this analysis?

**Comment by jublo** [comment]
Like, we can have a set of correct values to use for Hack font table releases, enabling us to safely ignore the development versions since the build will autofix them in the binaries?

**Comment by chrissimpkins** [comment]
What I am getting at is how difficult the changes will be from version to version of the releases.  The choice seems to be between work in XML and work in Python dictionaries.  The former I see as more difficult to parse, modify, and verify.  The latter is simple.  To stay within the Adobe tools, we must choose the former.  If we choose simplicity in the work, we could go with the fontTools library.  I think it depends upon how much up front and subsequent work will be necessary with the tool. 

Thoughts?

**Comment by jublo** [comment]
So fontTools can be instructed to open a TTF or an OTF, set some table flags and metrics, update the font name string and version strings, and write back to the binary?

**Comment by jublo** [comment]
I think both approaches (XML and fontTools) are equally easy to handle—at least with xml-sed, which can replace values inside the text representation by xPath key.

**Comment by chrissimpkins** [comment]
> So fontTools can be instructed to open a TTF or an OTF, set some table flags and metrics, update the font name string and version strings, and write back to the binary?

It can modify them for sure.  I need to confirm writes to both TTF and OTF formats.

**Comment by chrissimpkins** [comment]
xml-sed then introduces another dependency?

**Comment by jublo** [comment]
Would be dependent on `xml-coreutils` vs. be dependent on `fontTools`.

**Comment by jublo** [comment]
While I now see that we already rely on fonttools for `check_tools_subsetting`.

**Comment by chrissimpkins** [comment]
Python use would support windows users if we decide to move to an alternative to the Makefile approach

**Comment by jublo** [comment]
Good. The Makefile can easily be ported to an Ant, Grunt or Gulpfile.

**Comment by chrissimpkins** [comment]
Let me look into `fontTools` from a write standpoint in a bit more detail and we can discuss further.  I think that we need to know whether it supports both OTF and TTF writes.

**Comment by jublo** [comment]
@chrissimpkins So based on this discussion, we need a tool that takes a font binary, changes a specified value in a specified table, then writes back to another binary.

Suggested signature:
```
font-table-edit -in tmp/Hack-Regular.ttf -out tmp/Hack-Regular-updated.ttf -table "OS/2" update -key "ySubscriptYOffset" -value 491

font-table-edit -in tmp/Hack-Italic.otf -out tmp/Hack-Italic-updated.otf -table "NAME" update -nameid 123 -value "3.000"
```

**Comment by chrissimpkins** [comment]
I think that we could simplify it even further with a JSON or YAML settings file rather than the need for a long command line string.

**Comment by chrissimpkins** [comment]
This would allow us to make simple changes to the settings without the need to touch the script

**Comment by jublo** [comment]
However, I must admit that I’m nowhere near ready to code such in Python. I’ll have to leave that to you, @chrissimpkins …

**Comment by chrissimpkins** [comment]
Link to [Hack issue report 113](https://github.com/chrissimpkins/Hack/issues/113)

**Comment by chrissimpkins** [comment]
This is going to be possible in Python.  Will put together a test script this evening.

**Comment by chrissimpkins** [comment]
I can confirm that it works for both ttf and otf files.  I can modify the OT tables in memory and write out a new file.  There is no need to use XML.  I'll work on a script tomorrow.  Are you ok with YAML for the settings file?

**Comment by jublo** [comment]
Yep!

**Comment by chrissimpkins** [comment]
sounds good.  will let you know when it is ready

**Comment by chrissimpkins** [comment]
@jublo any idea where we are with the name tables?  Any problems that you identified there?

**Comment by jublo** [comment]
@chrissimpkins The font name carries over properly, the build-chain sets it prebuild already (Hack-dev). Same for the license etc.

**Comment by chrissimpkins** [comment]
thanks @jublo

**Comment by chrissimpkins** [comment]
@jublo The OT table modification script is now available at 

https://github.com/source-foundry/otmod

**Comment by chrissimpkins** [comment]
Any progress here?  Have some time to work on this with me this week?
